### ISSUE TITLE
Question about early stopping

### ISSUE THREAD

**Issue by Youskrpig** [talreiss/PANDA#4]
Hi, Thanks for your great work. I notice that " A weakness of the simple early-stopping approach, is the reliance on a hyper-parameter that may not generalize to new datasets. Although the optimal stopping epoch can be determined with a validation set containing anomalies, it is not available in our setting". I don' understand why simple early stopping with validation set is not available in OOC setting.   Validation set could be part of all normal images, and the change of validation loss is very small within seval epoches, then stop training. And i don't see the early stopping in code.

**Comment by NivC** [comment]
Hi Youskrpig,
Thank you for your interest in our work!

When discussing early stopping with a validation set, we mean a validation set that also contains anomalies. We want to optimize separation between normal images and anomalies, so we must have both in during validation. Yet, OCC assumes no anomalies are given at all before test time. Therefore, we did not find a validation set composed only of normal sample useful (but we are always open to new ideas :-) ). 

"And i don't see the early stopping in code." -  well, you can just set the stopping epoch you would like as an input parameter..

**Comment by Youskrpig** [comment]
Thanks for your reply. One more question in Section 3.2, "We can see that anomaly detection accuracy is correlated to having a large ratio between the distance of the anomalous samples to the center, and the distance between the normal samples and the center' What does this mean in fig3?  what does ---*---anomal/train  and ---*---nomal/train mean?  In code, i find center = torch.FloatTensor(feature_space).mean(dim=0) , Does it represent the ---*---nomal/train?
i'm really confused the distance curve? Could you please explain ?

**Comment by NivC** [comment]
"anomal/train" or "nomal/train" means the average distance between an anomalous (or normal) sample to the center, divided by the average distance on the normal train set.

The code does not generate this plot, but you can easily to it yourself.

Niv.

**Comment by Youskrpig** [comment]
Thanks for your reply. Appreciate！ One more question,  i run experiments with mvtec dataset, but there is almost no improvent, any suggestion?

**Comment by NivC** [comment]
As we mention in the paper, we only claim the method provides significant improvement on the larger datasets. Yet, we did see an improvement for some of the smaller ones, when training for many epochs. Generally, we want to sacle for a an even number of iterations (across datasets), so you may try to scale the number of epochs accordingly and see how it goes.

**Comment by Youskrpig** [comment]
Ok, i will try.

**Comment by szubing** [comment]
> Hi Youskrpig, Thank you for your interest in our work!
> 
> When discussing early stopping with a validation set, we mean a validation set that also contains anomalies. We want to optimize separation between normal images and anomalies, so we must have both in during validation. Yet, OCC assumes no anomalies are given at all before test time. Therefore, we did not find a validation set composed only of normal sample useful (but we are always open to new ideas :-) ).
> 
> "And i don't see the early stopping in code." - well, you can just set the stopping epoch you would like as an input parameter..

It seems the proposed Sample-wise early stopping (PANDA-SES) also need to use anomalies samples, right? So how it truly works exactly? I also don't find any code related to that.

**Comment by NivC** [comment]
Hi, the published code does not contain it as an option or so.
However, you can save checkpoints and evaluate Sample-wise early stopping as explained in the paper.

PANDA-SES does not use any anomalies during training but chooses a stopping epoch based on the single target sample being evaluated (which might be normal or anomalous). Please see the explanations in the paper.

Niv

**Comment by szubing** [comment]
ok, thanks for response
